Implement joint retargeting

GiulioRomualdi opened the pull request.

S-Dafarra commented:
Is there a plan to remove this from "draft"?

GiulioRomualdi commented:
In theory it should be ready. In practice I don't like some workarounds placed here and there to make everything working.

S-Dafarra commented on src/RetargetingHelper/src/Helper.cpp:
```diff
@@ -26,112 +26,328 @@ void RetargetingClient::convertYarpVectorPoseIntoTransform(const yarp::sig::Vect
-    m_rightHandSmoother->computeNextValues(m_rightHandTransformYarp);
-    convertYarpVectorPoseIntoTransform(m_rightHandSmoother->getPos(), m_rightHandTransform);
+
+        yInfo() << "ciao hand";
```
Is this necessary? :grin:

S-Dafarra commented on src/RetargetingHelper/src/Helper.cpp:
```diff
@@ -26,112 +26,328 @@ void RetargetingClient::convertYarpVectorPoseIntoTransform(const yarp::sig::Vect
-    if(desiredRightHandPose != nullptr)
-        m_rightHandTransformYarp = *desiredRightHandPose;
+        m_rightHand.smoother->computeNextValues(m_rightHand.yarpVector);
+        convertYarpVectorPoseIntoTransform(m_rightHand.smoother->getPos(), m_rightHandTransform);
```
You could try to write a function that does the same job for both hands and remove code duplications.

GiulioRomualdi replied:
Associated issue: https://github.com/dic-iit/bipedal-locomotion-controllers/issues/39

GiulioRomualdi replied:
In https://github.com/robotology/walking-controllers/pull/41/commits/2fdcdfaffe01255cd707c2ffd89a78128bd4615e, I tried to remove the duplicated code using lambda functions.

S-Dafarra commented on src/RetargetingHelper/src/Helper.cpp:
```diff
@@ -26,112 +26,328 @@ void RetargetingClient::convertYarpVectorPoseIntoTransform(const yarp::sig::Vect
+            auto desiredCoMHeight = m_comHeight.port.read(false);
+            if(desiredCoMHeight != nullptr)
+
+                m_comHeight.yarpVector(0) = ((*desiredCoMHeight)(2) - m_comHeightInputZero) * 0.5
```
Is the 0.5 a configuration parameter?

GiulioRomualdi replied:
Magic numbers :cry: :

GiulioRomualdi replied:
I remember that this heuristic was added as a workaround for an Xsens problem. As far as I understood the IK used to compute the `desiredCoMHeight` uses the model of the robot and the base directly coming from the Xsens. This means that the observer implemented in the Xsens generates a base displacement according to the human model. 
In other words, the CoM position is computed as follows:

![Untitled Diagram (7)](https://user-images.githubusercontent.com/16744101/80352181-c7079180-8873-11ea-87a5-16ec44fbbfbf.png)

The desired position of the CoM should depend on the desired position of the robot base (i.e. the human base pose scaled taking into account the difference between the robot and the human size)

The problem can be seen [here](https://www.youtube.com/watch?v=yELyMYkCyNE&feature=youtu.be&t=26). 

On the light of this we should decide, or we:
1. remove the CoM retargeting
2. fix the problem in the Xsens side

@S-Dafarra and @kouroshD

S-Dafarra replied:
I see. I think the problem here is that it is not completely clear which kind of information the walking controller is expecting. If the controller is expecting the height given the inertial frame it is considering, then the retargeting needs to know how this frame is located. In other words, the retargeting block needs to receive the base pose from the controller. I would argue that this is not the best choice since a reference quantity would depend on a measured quantity. 
The real question then is: which kind of reference the retargeting should generate?
Bonus question: how can we define this reference such that we can seamlessly switch from the XSens to the virtualizer?

Anyways, for the sake of this PR, I would remove it entirely, open an issue, and work on a new PR in order to add this feature, especially if this needs to be paired with a PR on ``walking-teleoperation``. I would also stress that the CoM retargeting was preventing the problem when raising the arms, so we should not let this fall in the oblivion.

GiulioRomualdi replied:
I agree so I will remove the CoM retargeting from the code

GiulioRomualdi replied:
Otherwise, I can keep the code as it is and disabling the CoM retargeting setting `m_useCoMHeightRetargeting = false`. This may help us in the future when we will reimplement the feature

S-Dafarra replied:
Ok, but either throw a warning, add a comment in the configuration, or whatever you prefer just to remind us of the status. 
If possible, I would move also all the magic numbers in the configuration files, but this would not be blocking.
Anyways, it is better to open an issue and start discussing on how to face this.

GiulioRomualdi replied:
The idea was to remove everything from the configuration file and disable the `CoMHeightRetargeting` adding 
```cpp
const m_useCoMHeightRetargeting = false
```
as an attribute of the class. If you don't like it I can remove everything

S-Dafarra commented on src/RetargetingHelper/src/Helper.cpp:
```diff
@@ -144,21 +360,104 @@ const iDynTree::Transform& RetargetingClient::rightHandTransform() const
 
+const iDynTree::VectorDynSize& RetargetingClient::jointValues() const
+{
+    return m_jointRetargetingValue;
```
How does the consumer class know if this output is meaningful or not? If I understood correctly, the hand retargeting and the joint retargeting cannot be enabled at the same time. Hence, how can you know if the value contained in this vector is garbage or not?

GiulioRomualdi replied:
we may return an `std::pair<bool, const iDynTree::VectorDynSize&>`. I don't really like it, but it could be a way to solve the problem.

S-Dafarra replied:
I agree, maybe adding a method like ``isJointRetargetingEnabled`` or whatever name you like? Do you think it may be an option?

S-Dafarra replied:
I don't remember if something changed here :thinking:

GiulioRomualdi replied:
Even if the joint retargeting is not enable the data are meaningful because `m_jointRetargetingValue` is reset using the initial position of the robot joints.

S-Dafarra replied:
Ok great!.

S-Dafarra commented on src/WalkingModule/src/Module.cpp:
```diff
@@ -322,6 +322,14 @@ bool WalkingModule::configure(yarp::os::ResourceFinder& rf)
+
+    m_stancePhaseMaxCounter = (int) std::round(stancePhaseTimeOutSeconds/m_dT);
+
+    yInfo() << "m_stancePhaseMaxCounter " << m_stancePhaseMaxCounter
```
Consider either to make this printout more detailed or to remove it.

GiulioRomualdi replied:
Removed in https://github.com/robotology/walking-controllers/pull/41/commits/9760be0556d233e940c1faba357e1d48f3f54cf8

S-Dafarra commented on src/WalkingModule/src/Module.cpp:
```diff
@@ -579,13 +594,30 @@ bool WalkingModule::updateModule()
+        }
+
+        // TODO remove me
+        yInfo() << "m_isStancePhase " << m_isStancePhase << " isStancePhaseStarting " << m_isStancePhaseStarting << " m_stancePhaseCounter " << m_stancePhaseCounter;
```
Can't we have this info from the planner?

GiulioRomualdi replied:
Moved in https://github.com/robotology/walking-controllers/pull/41/commits/9760be0556d233e940c1faba357e1d48f3f54cf8

S-Dafarra commented on src/WalkingModule/src/Module.cpp:
```diff
@@ -1239,6 +1289,22 @@ bool WalkingModule::setPlannerInput(double x, double y)
+        if(!m_isStancePhaseStarting)
+            m_stancePhaseCounter = m_stancePhaseMaxCounter;
+
+        m_isStancePhaseStarting = true;
```
Can't we create a function in the trajectory generator class asking for the phase value in a certain number of steps in the future?

GiulioRomualdi replied:
Done in https://github.com/robotology/walking-controllers/pull/41/commits/9760be0556d233e940c1faba357e1d48f3f54cf8

GiulioRomualdi commented:
I'm trying to merge `devel` (commit: ) on this branch. Unfortunately, there are some conflicts that we should resolve before proceeding. I will analyze every conflict so we can discuss in details:

- `RobotInterface/Helper.h` I introduced a `std::vector<bool>` called `m_dangerousJoints`. This vector should contain the list of the joints that are considered important for walking. I.e. when the flag is true the joint associated with the specific index is considered important for walking. If there is a problem in a _dangerous_ joint the controller stops. This was required because the wrist goes in hardware fault during retargeting. I would suggest adding all the parameters specific to the joints in a struct. For instance  
   ```cpp
   struct JointSpecification 
   {
       bool isInCompliantMode;
       bool isRequiredForLocomotion;
       int controlMode; // <----------- this might be useful If we want to control joints using a specific control mode. 
   };
   std::vector<JointSpecification> m_jointsSpecification;
   ```
    Thanks to this structure we can easily merge the code developed by @MiladShafiee and the one I implemented for the retargeting.
   Last but not least `m_dangerousJoint` is a terrible name. 

- `RobotInterface/src/Helper.cpp`. The code has to be updated accordingly to the design chosen for `RobotInterface/Helper.h`

-  `robotControl.ini` We should reintroduce the wrist retargeting. 

- `WholeBodyControllers/QPInverseKinematics.h` as already pointed by @S-Dafarra there are several indentation problems. Easy to solve (`Ctrl-x h TAB` and _les jeux sont faits_)

GiulioRomualdi commented on src/WalkingModule/app/robots/iCubGazeboV2_5/dcm_walking/joint_retargeting/jointRetargeting.ini:
```diff
@@ -0,0 +1,24 @@
+
+
+[JOINT_RETARGETING]
+retargeting_joint_list          ("neck_pitch", "neck_roll", "neck_yaw",
+                                 "torso_pitch", "torso_roll", "torso_yaw",
+                                 "l_shoulder_pitch", "l_shoulder_roll", "l_shoulder_yaw", "l_elbow", "l_wrist_prosup",
+                                 "r_shoulder_pitch", "r_shoulder_roll", "r_shoulder_yaw", "r_elbow", "r_wrist_prosup",
+                                 "l_hip_pitch", "l_hip_roll", "l_hip_yaw", "l_knee", "l_ankle_pitch", "l_ankle_roll",
+                                 "r_hip_pitch", "r_hip_roll", "r_hip_yaw", "r_knee", "r_ankle_pitch", "r_ankle_roll")
```
It is a subset of the joint controlled. For instance, in the simulation, we decided to retarget both the upper part and the lower part (just for testing). On the real robot, we enabled the retargeting only for the upper part

S-Dafarra replied:
Ok! So, is this the same ordering you have when getting the ``jointValues`` from the RetargetingHelper?

GiulioRomualdi replied:
Nope, the order is the one in `robotControl.ini`
i.e. 
```
 joints_list             ("neck_pitch", "neck_roll", "neck_yaw",
                         "l_shoulder_pitch", "l_shoulder_roll", "l_shoulder_yaw", "l_elbow", "l_wrist_prosup", "l_wrist_pitch", "l_wrist_yaw",
                         "r_shoulder_pitch", "r_shoulder_roll", "r_shoulder_yaw", "r_elbow", "r_wrist_prosup", "r_wrist_pitch", "r_wrist_yaw",
                         "l_hip_pitch", "l_hip_roll", "l_hip_yaw", "l_knee", "l_ankle_pitch", "l_ankle_roll",
                         "r_hip_pitch", "r_hip_roll", "r_hip_yaw", "r_knee", "r_ankle_pitch", "r_ankle_roll")
```
The joints contained in the `retargeting_joint_list` are the names of the joints coming from the Xsens module

S-Dafarra replied:
Ok, it's clear, thanks. Is it worth dropping a line somewhere?

GiulioRomualdi replied:
Ok I can add a comment in `robotControl.ini`

GiulioRomualdi replied:
Comment dropped in 5d0a4fda1eb186ca06f2c33f2f3074aea9c951c2

GiulioRomualdi commented on src/WalkingModule/app/robots/iCubGenova04/dcm_walking/common/pidParams.ini:
```diff
@@ -49,8 +49,8 @@ l_ankle_roll        2000.0       100.0         0.0
-r_ankle_roll       -2000.0      -100.0         0.0
+r_ankle_pitch      1000.0      100.0         0.0
+r_ankle_roll       2000.0      100.0         0.0
 
```
Given the current status of the robot, I think we can revert this

GiulioRomualdi replied:
Done in 02da2513df4056f2bc32fa2339f4ff79c2e6b60c

GiulioRomualdi commented on src/WalkingModule/app/robots/iCubGenova04/dcm_walking/hand_retargeting/robotControl.ini:
```diff
@@ -2,11 +2,16 @@ robot                   icub
-                        "l_shoulder_pitch", "l_shoulder_roll", "l_shoulder_yaw", "l_elbow", "l_wrist_prosup", "l_wrist_pitch", "l_wrist_yaw",
-                        "r_shoulder_pitch", "r_shoulder_roll", "r_shoulder_yaw", "r_elbow", "r_wrist_prosup", "r_wrist_pitch", "r_wrist_yaw",
+                        "l_shoulder_pitch", "l_shoulder_roll", "l_shoulder_yaw", "l_elbow", "l_wrist_prosup",
+                        "r_shoulder_pitch", "r_shoulder_roll", "r_shoulder_yaw", "r_elbow", "r_wrist_prosup",
```
There was a hardware problem related to the wrist. I think we can renable it.

GiulioRomualdi commented:
Another thing I really don't like with the solution I proposed is how the phases are handled. 
For instance, in the main module, the time is advanced only if we are not in the approaching phase. 
https://github.com/robotology/walking-controllers/blob/8c3640212919a689249ead8bc2c44ca344daf5c5/src/WalkingModule/src/Module.cpp#L845-L853
This new phase has been introduced with this PR and it is required to smoothing the trajectory coming from the `xsens` before running the controller. However, given the current implementation, there is an asymmetry between the `approaching phase` and the other phases. Indeed the other phases are handled using an `enum class` (check [here](https://github.com/robotology/walking-controllers/blob/feature/joint_retargeting/src/WalkingModule/include/WalkingControllers/WalkingModule/Module.h#L63)) while the `approaching phase` is only a simple boolean. 

Another terrible design choice is how the `stance phase` is managed. Given our current infrastructure, we are not able to teleoperate the robot while walking (this is a pitty I know but right now we have this limitation). To overcome this problem with decided to use a different set of weights for the walking and the stance phases in the QP optimization problem. 
Furthermore before switching to the `stance phase`, we decided to wait some seconds after the desired trajectory is ended  :cry:.
At this point, you may think: what the hell is he doing? Well, wait to hear this. I managed this delay with a counter directly inside the module.

Here the code where the counter is used:
https://github.com/robotology/walking-controllers/blob/8c3640212919a689249ead8bc2c44ca344daf5c5/src/WalkingModule/src/Module.cpp#L598-L608

The walking module is becoming bigger and bigger. When I developed the first version of the application, I didn't keep in mind that new features were required in the future (for instance Torque control, push recovery, retargeting). Each feature is different from the others. Probably in the future, we should split the logic of the module into submodules (or in blocks). Each block will have a common interface, then the final user (e.g. @S-Dafarra, @MiladShafiee @kouroshD) will combine the blocks in their own application. 

For the time being, we can try to implement a simple state machine for handling the switching between the different phases.

GiulioRomualdi commented:
As discussed with @S-Dafarra I updated the code. In details:
- https://github.com/robotology/walking-controllers/pull/41/commits/4a79c51f6a7f06980b2bed1eceeba5c777536758 I renamed `m_dangerousJoints` in `m_isGoodTrackingRequired`. I also introduced the two new methods for handling the joints in compliant mode.
- https://github.com/robotology/walking-controllers/pull/41/commits/21fb7bdf8cd9159e790e769bf54e5e0088b16711 I moved the logic behind the `stancePhase` inside `TrajectoryGenerator` class.

S-Dafarra commented:
I love the name ``m_isGoodTrackingRequired`` :grin:

GiulioRomualdi commented on src/RetargetingHelper/include/WalkingControllers/RetargetingHelper/Helper.h:
```diff
@@ -34,26 +36,44 @@ namespace WalkingControllers
+            yarp::os::BufferedPort<yarp::sig::Vector> port;
+            double smoothingTimeInApproaching;
+            double smoothingTimeInWalking;
+        } retargetingElement;
```
Done in f0eccade2f144e25beb621ae75610685e5678ba9

GiulioRomualdi commented on src/WalkingModule/src/Module.cpp:
```diff
@@ -378,9 +386,7 @@ bool WalkingModule::solveQPIK(const std::unique_ptr<WalkingQPIK>& solver, const
-    double threshold = 0.001;
-    bool stancePhase = iDynTree::toEigen(m_DCMVelocityDesired.front()).norm() < threshold;
-    solver->setPhase(stancePhase);
+    solver->setPhase(m_isStancePhase);
```
Now `m_isStancePhase` has been moved inside `TrajectoryGeneration` thanks  https://github.com/robotology/walking-controllers/pull/41/commits/9760be0556d233e940c1faba357e1d48f3f54cf8

GiulioRomualdi commented on src/WholeBodyControllers/include/WalkingControllers/WholeBodyControllers/QPInverseKinematics.h:
```diff
@@ -95,302 +96,339 @@ namespace WalkingControllers
-        const iDynTree::VectorDynSize& getLowerBound() const;
-    };
-};
+                                                                                       double m_kJointLimitsLowerBound; /**< Gain related to the the joint lower bound */
```
Fixed!

GiulioRomualdi commented on src/WholeBodyControllers/include/WalkingControllers/WholeBodyControllers/QPInverseKinematics.h:
```diff
@@ -95,302 +96,339 @@ namespace WalkingControllers
+                                                                                       const iDynTree::Twist& rightHandTwist);
+
+                                                                                       /**
+                                                                                       * Set the desired twist of both feet
```
Fixed!

GiulioRomualdi commented on src/WholeBodyControllers/src/QPInverseKinematics.cpp:
```diff
@@ -546,14 +677,38 @@ void WalkingQPIK::evaluateGradientVector()
+        // g = Weight * K_p * (regularizationTerm - jointPosition)
+        // Weight  and K_p are two diagonal matrices so their product can be also evaluated multiplying component-wise
+        // the elements of the vectors and then generating the diagonal matrix
+        gradient.tail(m_actuatedDOFs) += (-jointRegularizationGains.cwiseProduct(
+                                              iDynTree::toEigen(m_jointRegularizationWeightSmoother->getPos()))).asDiagonal()
+            * (regularizationTerm - jointPosition)
+            + (-jointRetargetingGains.cwiseProduct(iDynTree::toEigen(m_jointRetargetingWeightSmoother->getPos()))).asDiagonal()
+            * (jointRetargetingValues - jointPosition);
```
When you say splitting do you mean to store half of the computation in an `Eigen::VectorXd`? I

S-Dafarra replied:
Yes, you can use some intermediate buffers. Also if you have something like
``
a = (b*c) + (c*d);
``
you can split it in
``
a = (b*c);
a += (c*d);
``
For more info, https://eigen.tuxfamily.org/dox/TopicLazyEvaluation.html

GiulioRomualdi replied:
> Yes, you can use some intermediate buffers. Also if you have something like
> 
> ```
> a = (b*c) + (d*e);
> ```
> 
> you can split it in
> 
> ```
> a = (b*c);
> a += (d*e);
> ```
> 
> For more info, https://eigen.tuxfamily.org/dox/TopicLazyEvaluation.html

Shouldn't this reduce the performances? As far as I understood `a = (b*c) + (d*e)` Eigen will decide whether to evaluate it into a temporary variable. Thus if we are lucky no temporary variables are required. On the other hand 
```
a = (b*c);
a += (d*e);
```
will require the evaluation of the first part in the variable `a`. 

Probably this may solve our problems
```cpp
auto tmp1 = (b*c);
auto tmp2 = (d*e);
a = tmp1 + tmp2;
```
however, I am not a big fan of using `auto` with `Eigen` (you never know which kind of type are you using)

By the way, I completely agree that sometimes the readability is more important than the performances, so I can split the computation in two parts.

S-Dafarra replied:
> Shouldn't this reduce the performances? As far as I understood a = (b*c) + (d*e) Eigen will decide whether to evaluate it into a temporary variable. Thus if we are lucky no temporary variables are required.

Matrix multiplication is one of the cases where Eigen evaluates the sub-expression in a temporary object. Hence, I guess that the advantage of avoiding an assignment is eaten up by the need for allocating memory.

GiulioRomualdi replied:
Thanks for the explanation. 
f7442cff90563d0294c24214332117865ad41652 I split the computation as you suggested

GiulioRomualdi commented on src/WholeBodyControllers/src/QPInverseKinematics.cpp:
```diff
@@ -43,6 +43,14 @@ bool WalkingQPIK::initialize(const yarp::os::Searchable &config,
     m_enableHandRetargeting = config.check("use_hand_retargeting", yarp::os::Value(false)).asBool();
+    m_enableJointRetargeting = config.check("use_joint_retargeting", yarp::os::Value(false)).asBool();
+
+    if(m_enableHandRetargeting && m_enableJointRetargeting)
```
I did this for splitting the `RetargetingHelper` from the `QPInverseKinematics`. Probably we can remove the boolean and add an `enum` for changing the behavior of the solver

S-Dafarra replied:
As you prefer.

GiulioRomualdi replied:
In 787ac48394c05a60c5bf9edaafa0c51e2d43a857 `m_enableHandRetargeting` and `m_enableJointRetargeting` are replaced by an `enum class`

S-Dafarra commented on src/WalkingModule/src/Module.cpp:
```diff
@@ -499,10 +509,17 @@ bool WalkingModule::updateModule()
+                                          * m_FKSolver->getLeftHandToWorldTransform(),
+                                          m_FKSolver->getHeadToWorldTransform().inverse()
+                                          * m_FKSolver->getRightHandToWorldTransform(),
+                                          m_robotControlHelper->getJointPosition(),
```
Can you get also the joint values from the ``m_FKSolver`` to avoid the problem we had about the discrepancy? In principle, you could pass directly the ``m_FKSolver`` object.

GiulioRomualdi replied:
> Can you get also the joint values from the `m_FKSolver` to avoid the problem we had about the discrepancy?

Yes, actually you spot the bug already discovered by @MiladShafiee.  I'm updating the code

> In principle, you could pass directly the `m_FKSolver` object.

Yes it would be nice, but I would avoid to do it within this PR

GiulioRomualdi replied:
> Yes, actually you spot the bug already discovered by @MiladShafiee. I'm updating the code

Actually no. The bug is in 
```cpp
 ok &= solver->setRobotState(m_robotControlHelper->getJointPosition(),                            
                                 m_FKSolver->getLeftFootToWorldTransform(),
                                 m_FKSolver->getRightFootToWorldTransform(),
                                 m_FKSolver->getLeftHandToWorldTransform(),
                                 m_FKSolver->getRightHandToWorldTransform(),
                                 m_FKSolver->getNeckOrientation(),
                                 m_FKSolver->getCoMPosition());
```
The one you point is correct. The `retargetingClient` required the current position of the joints

S-Dafarra replied:
Actually, I suggested to pass the joint positions from the ``FKSolver`` to avoid exactly this kind of problems. Otherwise, how can I understand if this method needs the measured or the desired values? It seems very confusing.

S-Dafarra replied:
Great! Can you pass it as a ``const`` object or there are issues? Can the same be done also for the ``setRobotState``? (Just wondering)

GiulioRomualdi replied:
> Can the same be done also for the setRobotState? (Just wondering)

Here you have: ab9da12a431fbf7b77bd0fce5ae2d47d82556525ab9da12a431fbf7b77bd0fce5ae2d47d82556525

GiulioRomualdi commented on src/RobotInterface/src/Helper.cpp:
```diff
@@ -213,32 +213,40 @@ bool RobotInterface::configureRobot(const yarp::os::Searchable& config)
         else
         {
-            m_isJointModeStiffVector.at(i)=yarp::dev::InteractionModeEnum::VOCAB_IM_COMPLIANT;
+            m_jointInteractionMode[i] = yarp::dev::InteractionModeEnum::VOCAB_IM_COMPLIANT;
```
Yes I can print the warning only once at the beginning.

GiulioRomualdi replied:
Done in a4621db3a494f27670b39fbab73f763d4fe67025

GiulioRomualdi commented on src/TrajectoryPlanner/include/WalkingControllers/TrajectoryPlanner/TrajectoryGenerator.h:
```diff
@@ -226,6 +227,15 @@ namespace WalkingControllers
+         * stance foot during SS, 1 in case of swing foot)
+         * @return true/false in case of success/failure.
+         */
+        bool getIsStancePhase(std::vector<bool>& isStancePhase);
```
Thanks! Fixed in 8026c26afcf2c23bbdad12e1f8556614cf194469

GiulioRomualdi commented on src/RetargetingHelper/include/WalkingControllers/RetargetingHelper/Helper.h:
```diff
@@ -62,26 +83,38 @@ namespace WalkingControllers
                         const std::string &name,
-                        const double &period);
+                        const double &period,
+                        const std::vector<std::string>& controlledJointsName);
```
Done in https://github.com/robotology/walking-controllers/pull/41/commits/0acf4fa3f1772b3986c4af5f423cc1aaa5649bac

S-Dafarra commented on src/WholeBodyControllers/include/WalkingControllers/WholeBodyControllers/QPInverseKinematics.h:
```diff
@@ -280,6 +304,21 @@ namespace WalkingControllers
+        void setDesiredHandsTwist(const iDynTree::Twist& leftHandTwist,
+                                  const iDynTree::Twist& rightHandTwist);
+
+        /**
+         * Set the desired twist of both feet
+         * @param jointPosition contain the desired joint position used in the retargeting.
+         * @return true/false in case of success/failure.
+         */
+        bool setDesiredRetargetingJoint(const iDynTree::VectorDynSize& jointPosition);
```
The documentation seems wrong.

GiulioRomualdi replied:
Done in c2828ba194b295b916502f36e776a0efe488979e

S-Dafarra commented on src/WalkingModule/src/Module.cpp:
```diff
@@ -1274,6 +1303,8 @@ bool WalkingModule::setPlannerInput(double x, double y)
 
     m_newTrajectoryRequired = true;
 
+    double treshold = 0.01;
+
```
Is this needed?

GiulioRomualdi replied:
Fixed in a9a2591e7e1e2a6a93e5ff5a08cfdcb51d9b3003

S-Dafarra commented on src/RetargetingHelper/include/WalkingControllers/RetargetingHelper/Helper.h:
```diff
@@ -103,10 +135,41 @@ namespace WalkingControllers
         void setRobotBaseOrientation(const iDynTree::Rotation& rotation);
+
+
+        void setPhase(bool isStancePhase);
```
Friendly ping

GiulioRomualdi replied:
Done in d86a5627f5cf62d6dc98c004db020986e5bfc564

S-Dafarra commented on src/RetargetingHelper/include/WalkingControllers/RetargetingHelper/Helper.h:
```diff
@@ -103,10 +135,41 @@ namespace WalkingControllers
+        void setPhase(bool isStancePhase);
+
+        /**
+         * Start the approaching phase
```
Friendly ping

GiulioRomualdi replied:
Done in 4650260fd5f3a502e37933377e99db654bb6deaf

S-Dafarra commented on src/WholeBodyControllers/src/QPInverseKinematics.cpp:
```diff
@@ -546,14 +677,38 @@ void WalkingQPIK::evaluateGradientVector()
+        // g = Weight * K_p * (regularizationTerm - jointPosition)
+        // Weight  and K_p are two diagonal matrices so their product can be also evaluated multiplying component-wise
+        // the elements of the vectors and then generating the diagonal matrix
+        gradient.tail(m_actuatedDOFs) += (-jointRegularizationWeights.cwiseProduct(jointRegularizationGains)).asDiagonal()
```
Friendly ping.

GiulioRomualdi replied:
Done in 220c4764d2cb52eb2241dd3c99d674ee4114beb2

S-Dafarra commented:
By looking at the file, I noticed there were still some old comments. Probably some of them have been solved already. Feel free to resolve those you think do not apply anymore.

GiulioRomualdi commented on src/WalkingModule/app/robots/iCubGazeboV2_5/dcm_walking/joint_retargeting/qpInverseKinematics.ini:
```diff
@@ -0,0 +1,78 @@
+torso_weight_stance             0.0
+additional_rotation             ((0.0 0.0 1.0),(1.0 0.0 0.0),(0.0 1.0 0.0))
+
+## Regularization
```
Done in 22992f9b905c784fdb8208059bce24879cbce286

GiulioRomualdi commented on src/KinDynWrapper/src/Wrapper.cpp:
```diff
@@ -517,3 +517,13 @@ bool WalkingFK::getCoMJacobian(iDynTree::MatrixDynSize &jacobian)
+
+iDynTree::VectorDynSize WalkingFK::getJointPos()
+{
+    iDynTree::VectorDynSize jointPos;
```
Done in fde9abd53a560d6b030c1646df3609792214a320

GiulioRomualdi commented:
@S-Dafarra, I think I addressed all your comments.

GiulioRomualdi commented:
I'm going to merge the PR. 
The latest commit tested on the robot is https://github.com/robotology/walking-controllers/pull/41/commits/169988c9d84b71e662a6f080dfa9442fab91a386. In case of problems we can switch back to that commit
